third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -1818,6 +1886,68 @@ impl Frame {
+                let output_ty = resolver.create_struct_type(&info.definition_struct_type);
+                Self::verify_pack(interpreter, info.field_count, field_tys, output_ty)?;
+            },
+            Bytecode::PackVariantGeneric(idx) => {
+                let info = resolver.get_struct_variant_instantiation_at(*idx);
+                let output_ty = ty_cache
+                    .get_struct_variant_type(*idx, resolver, ty_args)?
+                    .0
+                    .clone();
+                let args_ty = ty_cache.get_struct_variant_fields_types(*idx, resolver, ty_args)?;
+                Self::verify_pack(
+                    interpreter,
+                    info.field_count,
+                    args_ty.iter().map(|(ty, _)| ty),
+                    output_ty,
+                )?;
+            },
+            Bytecode::UnpackVariant(idx) => {
+                let info = resolver.get_struct_variant_at(*idx);
+                let expected_struct_ty = resolver.create_struct_type(&info.definition_struct_type);
+                let actual_struct_ty = interpreter.operand_stack.pop_ty()?;
+                actual_struct_ty.paranoid_check_eq(&expected_struct_ty)?;
+                for (_name, ty) in info
+                    .definition_struct_type
+                    .fields(Some(info.variant))
+                    .iter()
+                {
+                    interpreter.operand_stack.push_ty(ty.clone())?;
+                }
+            },
+            Bytecode::UnpackVariantGeneric(idx) => {
+                let expected_struct_type =
+                    ty_cache.get_struct_variant_type(*idx, resolver, ty_args)?.0;
+                let actual_struct_type = interpreter.operand_stack.pop_ty()?;
+                actual_struct_type.paranoid_check_eq(expected_struct_type)?;
+                let struct_fields_types =
+                    ty_cache.get_struct_variant_fields_types(*idx, resolver, ty_args)?;
+                for (ty, _) in struct_fields_types {
+                    interpreter.operand_stack.push_ty(ty.clone())?;
+                }
+            },
+            Bytecode::TestVariant(idx) => {
+                let info = resolver.get_struct_variant_at(*idx);
+                let expected_struct_ty = resolver.create_struct_type(&info.definition_struct_type);
+                let actual_struct_ty = interpreter.operand_stack.pop_ty()?;
+                actual_struct_ty.paranoid_check_ref_eq(&expected_struct_ty, false)?;
+                interpreter.operand_stack.push_ty(Type::Bool)?;
+            },
+            Bytecode::TestVariantGeneric(idx) => {
+                let expected_struct_ty =
+                    ty_cache.get_struct_variant_type(*idx, resolver, ty_args)?.0;
+                let actual_struct_ty = interpreter.operand_stack.pop_ty()?;
+                actual_struct_ty.paranoid_check_ref_eq(expected_struct_ty, false)?;
+                interpreter.operand_stack.push_ty(Type::Bool)?;
```

Comment:
Please use type builder: `ty_builder.create_bool_ty()`. We do not have proper type encapsulation so that only type builder can create them because then we add ugly `TypeImpl`s and I did not get there, but we should always create types in a single way using the builder